smartconfig/entry.py

```diff
@@ -0,0 +1,138 @@
+
+from smartconfig import _registry
+from smartconfig._registry import configuration_for_module, global_configuration, lookup_global_configuration
+from smartconfig.exceptions import ConfigurationKeyError, InvalidOperation, PathConflict
+
+
+class _ConfigEntryMeta(type):
+    """
+    Metaclass used to define special ConfigEntry behaviors.
+
+    Note: Using this metaclass outside of the library is currently not supported.
+
+    The lookup of attributes can be customized by subclassing this metaclass and changing `_get_attribute_path`.
+    """
+
+    def __getattribute__(cls, name: str) -> Optional[Any]:
+        """
+        Look up the attribute through the configuration system.
+
+        If the attribute name starts with `_`, normal lookup is done, otherwise _registry.global_configuration is used.
+
+        Args:
+            name: Attribute to lookup
+
+        Raises:
+            ConfigurationKeyError: The attribute doesn't exist.
+        """
+        # Use the normal lookup for attribute starting with `_`
+        if name.startswith('_'):
+            return super().__getattribute__(name)
+
+        path, attribute = cls._get_attribute_path(name)
+
+        return lookup_global_configuration(path, attribute)
+
+    def __new__(cls, name: str, bases: Tuple[type, ...], dict_: Dict[str, Any], path: Optional[str] = None) -> type:
+        """
+        Add the `__defined_attributes` and `__path_override` attributes to the new entry.
+
+        Args:
+            path: Custom path used for this entry instead of the module name.
+        """
+        names = chain(dict_.keys(), dict_.get('__annotations__', {}).keys())
+        dict_[f"{cls.__name__}__defined_attributes"] = {name for name in names if not name.startswith('_')}
+
+        dict_[f"{cls.__name__}__path_override"] = path
+
+        return super().__new__(cls, name, bases, dict_)
+
+    def _register_entry(cls) -> None:
+        """Set the `__path` attribute and register the entry."""
+        cls.__path = cls.__path_override or cls.__module__
+        if cls.__path in _registry.configuration_for_module:
+            raise PathConflict(f"An entry at {cls.__path!r} already exists.")  # TODO: Add an FAQ link.
+
+        configuration = {
+            key: value for key, value in cls.__dict__.items() if not key.startswith('_')
+        }
+
+        if cls.__path not in global_configuration:
+            global_configuration[cls.__path] = configuration
+        # We already have some overrides for this path.
+        else:
+            for key, value in configuration.items():
+                path, key = cls._get_attribute_path(key)
+
+                if path not in global_configuration:
+                    global_configuration[path] = {}
+
+                # We only write values that aren't already defined.
+                if key not in global_configuration[path]:
+                    global_configuration[path][key] = value
+
+        configuration_for_module[cls.__path] = cls
```

Comment:
The idea here would be to have it for user defined code. For example, the extension library will need to have a list of all the defined entries, I think it is pretty useful to store them here.

Reply:
This currently isn't exposed as a public API, and doing so will probably be bad since tampering could easily break things. If you think exposing this is useful for extensibility, then the way to go about it would be to expose a read-only view for the dictionary (or just the values if the paths aren't useful).

In any case, if it's going to be storing the classes, it may be better to use a weak reference dictionary.

Reply:
I moved to a set since we can use `ConfigEntry.__subclasses__` in order to get the currently loaded entries. At worse, we can always create a new API later.